### ISSUE TITLE
Inaccurate key frame seeking on MP4 format

### ISSUE THREAD

**Issue by msg7086** [Nevcairiel/LAVFilters#232]
I'm using latest MPC-HC with LAVFilters, and I recently upgraded its internal LAVFilters to a later version and notice that it no longer jump to the key frames, but 2 frames behind it.

For example, the same video muxed in MKV reports a key frame at 7823, while when mixed in MP4, it reports one at 7821, which is 2 frames behind the correct position.

I did a quick bisect on releases, and confirmed 0.71 was good and 0.72 started to cause the issue. I couldn't track it further down to whether LAV is at fault or an upgrade of ffmpeg caused that.

The file I used for testing is x264 8-bit video, muxed to MP4 by l-smash. Another x265 10-bit video was found to have 4 frames difference, eg. reports one at 100 while the actual key frame is at 104.

Keeping digging down. I tried to mix LAVSplitter 0.71 with LAVVideo 0.72. I noticed that, while LAVSplitter reports 7821 as the key frame, LAVVideo indeed decodes and returns 7823. LAVVideo 0.72 behaves the same as LAVVideo 0.71.

I'm kind of confused now. What's the expect behavior of LAVSplitter here? Sounds to me like the new splitter fixed something and caused something else to break?

**Comment by msg7086** [comment]
TL;DR, suppose a key frame is at 100 for an x264 video.

* LAVSplitter 0.71

    * MKV
        Jumps to 100. Renders 100. Steps to 101.

    * MP4
        Jumps to 98. Renders 100. Steps to 101.

* LAVSplitter 0.72

    * MKV
        Jumps to 100. Renders 100. Steps to 101.

    * MP4
        Jumps to 98. Renders 98 (Second to last frame from previous GOP). Steps to 99 (Last frame from previous GOP).

**Comment by clsid2** [comment]
It is a bug in the ffmpeg code that is used by LAV Filters. I gave Nevcairiel a sample file for this issue a few weeks ago, but he hasn't had time to investigate it yet.

**Comment by Nevcairiel** [comment]
As far as I can tell, the timestamps that LAV reports as key frames haven't actually changed, what has changed is the seek behavior. Due to some unfortunate design decisions, the timestamps in the MP4 keyframe index are in DTS (Decoding Timestamps), while the seek function in more recent versions expects PTS (Presentation Timestamps). PTS generally makes more sense for seeking, since its the actual time the frame will appears on screen -  but unfortunately it mismatches the index.

In any case, I've tried to apply a fix in LAV Splitter directly that tries to offset the timestamps from DTS to PTS, reversing the logic that the FFmpeg mov/mp4 demuxer uses. So now the timestamps of the keyframes actually do change, and perhaps get closer to what you would get from a MKV, and it should fix seeking too.

**Comment by clsid2** [comment]
Similar issue here:
https://github.com/clsid2/mpc-hc/issues/26

The file has a keyframe at 05:55 and 06:04. If you seek to the 05:55 keyframe, then the video hangs until the next keyframe.

**Comment by Nevcairiel** [comment]
Seeking works perfectly on that sample. Seek to 3546459583 (354.65s), and the next frame output has that exact timestamp. It just has wrong keyframe flags, I reckon.
